frappe/handler.py

```diff
@@ -230,22 +218,13 @@ def run_doc_method(method, docs=None, dt=None, dn=None, arg=None, args=None):
 		doc._original_modified = doc.modified
 		doc.check_if_latest()
 
-	if not doc or not doc.has_permission("read"):
-		throw_permission_error()
-
-	try:
-		args = json.loads(args)
-	except ValueError:
-		args = args
-
-	method_obj = getattr(doc, method)
-	fn = getattr(method_obj, '__func__', method_obj)
-	is_whitelisted(fn)
-	is_valid_http_method(fn)
+	if not doc or not doc.has_permission():
+		frappe.throw(_("Not permitted"), frappe.PermissionError)
 
-	fnargs = inspect.getfullargspec(method_obj).args
+	doc.is_whitelisted(method)
```

Comment:
Shouldn't this be checked before `has_permission`?

Reply:
Both are validations. Ideally, we expect the user to not make any of these mistakes. Despite that, I guess it's a matter of assumption to decide which mistake the user is more likely to make.

In this case, I have just done what was simpler and let the earlier logic to check permission be as-is. That way, by the time we come to `doc.is_whitelisted`, we know that `doc` is not falsy.

Reply:
PS: maybe I should show a better error message for when `doc` doesn't exist?